### PR TITLE
Chainsafe Audit Changes

### PR DESCRIPTION
**Pull request by chuddster**
- withdraw from yearn vault correctly in upgradeYearnVault

### CODE REVIEW

File: contracts/libraries/Vault.sol [hunk 1]
```diff
@@ -60,6 +60,8 @@ library Vault {
         uint128 totalPending;
         // Amount locked for scheduled withdrawals;
         uint128 queuedWithdrawShares;
+        // Amount locked for scheduled withdrawals last week;
+        uint128 lastQueuedWithdrawAmount;
```
Comment (kenchangh) [inline contracts/libraries/Vault.sol:64]: We can't append `lastQueuedWithdrawAmount` into the `VaultState` because it will break the storage layout. So we need to append it into RibbonThetaVaultStorageV2 as a variable

```solidity
contract RibbonThetaVaultStorageV2 {
  uint256 lastQueuedWithdrawAmount;
}
```

File: contracts/libraries/VaultLifecycleSTETH.sol [hunk 2]
```diff
@@ -322,7 +327,8 @@ library VaultLifecycleSTETH {
                 withdrawAmount.sub(yieldTokenBalance)
             );
 
-        transferAsset(payable(recipient), underlyingTokensToWithdraw);
+        IWETH(weth).deposit{value: underlyingTokensToWithdraw}();
+        IERC20(weth).safeTransfer(recipient, underlyingTokensToWithdraw);
```
Comment (kenchangh) [inline contracts/libraries/VaultLifecycleSTETH.sol:331]: What is the rationale behind this change?

File: contracts/vaults/BaseVaults/base/RibbonVault.sol [hunk 3]
```diff
@@ -591,11 +591,22 @@ contract RibbonVault is
...
+            uint256(vaultState.lastQueuedWithdrawAmount);
+
         // Take management / performance fee from previous round and deduct
         lockedBalance = _lockedBalance.sub(
-            _collectVaultFees(_lockedBalance.add(queuedWithdrawAmount))
+            _collectVaultFees(
+                _lockedBalance.add(
+                    queuedWithdrawAmount > lastQueuedWithdrawAmount
+                        ? queuedWithdrawAmount.sub(lastQueuedWithdrawAmount)
+                        : 0
```
Comment (kenchangh) [inline contracts/vaults/BaseVaults/base/RibbonVault.sol:603]: Probably makes sense to make this a variable like `withdrawAmountDiff` to help with readability

File: contracts/libraries/VaultLifecycleSTETH.sol (hunk 2 above)
Comment (chuddster) [reply contracts/libraries/VaultLifecycleSTETH.sol:331]: if we use curve, the burner just does swapExactTokensForTokens. so figure it will be easier later down the line for RBN buyback to just have all ERC-20 tokens

File: contracts/libraries/Vault.sol (hunk 1 above)
Comment (chuddster) [reply contracts/libraries/Vault.sol]: addressed this

File: contracts/storage/RibbonThetaSTETHVaultStorage.sol [hunk 4]
```diff
@@ -22,7 +22,7 @@ abstract contract RibbonThetaSTETHVaultStorageV1 {
 
 abstract contract RibbonThetaSTETHVaultStorageV2 {
     // Amount locked for scheduled withdrawals last week;
-    uint128 public lastQueuedWithdrawAmount;
+    uint256 public lastQueuedWithdrawAmount;
```
Comment (kenchangh) [inline contracts/storage/RibbonThetaSTETHVaultStorage.sol:25]: Changed it to uint256 to save gas

**Review by kenchangh — APPROVED** [review]
